allocator.go

```diff
@@ -114,40 +118,43 @@ func (a *Allocator) adhereToNodeResourcesConstraint() {
 	}
 }
 
-func (a *Allocator) adhereToNodeTagsConstraint() {
+func (a *Allocator) adhereToNodeTags() {
 	for _, r := range a.ranges {
-		nodeThatAreUnAssignable := make([]cpsatsolver.Literal, 0, len(a.nodes))
+		nodeThatAreNonassignable := make([]cpsatsolver.Literal, 0, len(a.nodes))
 		for _, n := range a.nodes {
 			if !rangeTagsAreSubsetOfNodeTags(r.tags, n.tags) {
-				nodeThatAreUnAssignable = append(nodeThatAreUnAssignable, a.literal(r, n))
+				nodeThatAreNonassignable = append(nodeThatAreNonassignable, a.literal(r, n))
 			}
 		}
 		a.model.AddConstraints(
-			cpsatsolver.NewAllowedLiteralAssignmentsConstraint(
-				nodeThatAreUnAssignable,
-				[][]bool{
-					make([]bool, len(nodeThatAreUnAssignable)),
-				}),
+			cpsatsolver.NewExactlyKConstraint(0, nodeThatAreNonassignable...),
 		)
 	}
 }
 
-func rangeTagsAreSubsetOfNodeTags(tOne []string, tTwo []string) bool {
-	for _, rangeTag := range tOne {
-		for tagIndex, nodeTag := range tTwo {
+func rangeTagsAreSubsetOfNodeTags(rangeTags []string, nodeTags []string) bool {
+	for _, rangeTag := range rangeTags {
+		foundMatch := false
+		for _, nodeTag := range nodeTags {
 			if rangeTag == nodeTag {
+				foundMatch = true
 				break
-			} else if tagIndex == len(tTwo)-1 {
-				return false
 			}
 		}
+		if !foundMatch {
+			return false
+		}
 	}
 	return true
 }
 
-func (a *Allocator) Allocate(constraintOptions ...ConstraintOption) (ok bool, assignments map[RangeID][]NodeID) {
```

Comment:
What you had before was more right, options only make sense during initialization should be passed in during initialization. Right now it's possible to `a := allocator.New(); a.Allocate(); a.WithMatchingTagConstraint()`/ How about including these variadic arguments as part of `allocator.New()`?

Reply:
Agreed that `new` would be the easiest, it helps to do this there too instead of creating an `opts` struct initialized with each field equalling `false`.

Reply:
Pushed the change, @irfansharif please let me know if this is better.

Reply:
Ah right zero-values, good call, added this into the `New()` method.